src/pp/if_parser.rs

```diff
@@ -0,0 +1,436 @@
+    pub fn new(
+        tokens: Vec<Token>,
+        defines: &'macros HashMap<String, Rc<Define>>,
+        location: Location,
+        parsing_if: bool,
+    ) -> Self {
+        IfParser {
+            lexer: IfLexer {
+                tokens: tokens.into_iter(),
+                defines,
+            },
+            macro_processor: MacroProcessor::default(),
+            location,
+
+            parsing_if,
+            carry: None,
+        }
+    }
+
+    fn raw_next(&mut self) -> Option<Token> {
+        self.carry
+            .take()
+            .or_else(|| self.macro_processor.step(&mut self.lexer).ok())
+    }
+
+    pub fn raw_peek(&mut self) -> Option<Token> {
+        self.carry.clone().or_else(|| {
+            let token = self.macro_processor.step(&mut self.lexer).ok();
+
+            self.carry = token.clone();
+
+            token
+        })
+    }
+
+    fn next(&mut self) -> Step<Option<Token>> {
+        let token = match self.raw_next() {
+            Some(t) => t,
+            None => return Ok(None),
+        };
+
+        Ok(match token.value {
+            TokenValue::Ident(ref name) if name != "defined" => {
+                match self.add_define(name, token.location)? {
+                    Some(t) => Some(t),
+                    None => self.next()?,
+                }
+            }
+            _ => Some(token),
+        })
+    }
+
+    fn peek(&mut self) -> Step<Option<Token>> {
+        let token = match self.raw_peek() {
+            Some(t) => t,
+            None => return Ok(None),
+        };
+
+        Ok(match token.value {
+            TokenValue::Ident(ref name) if name != "defined" => {
+                self.raw_next();
+
+                match self.add_define(name, token.location)? {
+                    Some(t) => {
+                        self.carry = Some(t);
+                        self.carry.clone()
+                    }
+                    None => self.peek()?,
+                }
+            }
+            _ => Some(token),
+        })
+    }
+
+    fn expect_raw_next(&mut self) -> Step<Token> {
+        let val = self.raw_next();
+        val.ok_or(StepExit::Error((
+            PreprocessorError::UnexpectedEndOfInput,
+            self.location,
+        )))
+    }
+
+    fn expect_next(&mut self) -> Step<Token> {
+        let val = self.next()?;
+        val.ok_or(StepExit::Error((
+            PreprocessorError::UnexpectedEndOfInput,
+            self.location,
+        )))
+    }
+
+    fn expect_peek(&mut self) -> Step<Token> {
+        let val = self.peek()?;
+        val.ok_or(StepExit::Error((
+            PreprocessorError::UnexpectedEndOfInput,
+            self.location,
+        )))
+    }
+
+    fn add_define(&mut self, name: &str, location: Location) -> Step<Option<Token>> {
+        if self
+            .macro_processor
+            .start_define_invocation(name, location, &mut self.lexer)?
+        {
+            Ok(None)
+        } else if self.parsing_if {
```

Comment:
nit: add a comment that explains what this handles?

Reply:
There's already a comment in `Self::new` explaining the behavior